### DISCUSSION TITLE
Patterns for passing dynamic options to directives

### DISCUSSION THREAD

**Discussion by luisherranz** [WordPress/gutenberg#52973] [Interactivity API]
Directives that instead of a reference receive a list of options could still accept a reference that returns those options when they need to be modified dynamically.

For example, a directive `data-wp-example` that has a couple of options, like this:

```html
<div data-wp-example='{ "option1": "value", "option2": "value" }'>...</div>
```

Could also accept a reference that returns those options dynamically:

```html
<div data-wp-example="options.namespace.someExample">...</div>
```

```js
store({
  options: {
    namespace: {
      someExample: ({ state }) => {
        if (state.namespace.someValue === true)
          return { option1: "value1", option2: "value2" };
        else return { option1: "value3", option2: "value4" };
      },
    },
  },
});
```

This pattern could also be applied _per option_.

```html
<div
  data-wp-example='{ "option1": "value", "option2": "ref:options.namespace.someOption2" }'
>
  ...
</div>
```

```js
store({
  options: {
    namespace: {
      someOption2: ({ state }) => {
        if (state.namespace.someValue === true) return "value2";
        else return "value4";
      },
    },
  },
});
```

Whenever it makes sense, instead of the value of the option (or apart from it), the functionality itself could be delegated. For example, imagine this directive that does client-side navigation with options for `prefetch` and `scroll`:

```html
<a data-wp-navigation-link='{ "prefetch": "hover", "scroll": true }'> ... </a>
```

Instead of passing the option to `scroll` dynamically, the reference could handle the scroll logic itself.

```html
<a
  data-wp-navigation-link='{ "prefetch": "hover", "scroll": "ref:actions.namescape.scrollToSomeElement }'
>
  ...
</a>
```

```js
store({
  actions: {
    namespace: {
      scrollToSomeElement: ({ ref }) => {
        ref.closest(".some-element").scrollIntoView();
        return false;
      },
    },
  },
});
```

The `ref:` prefix is not enforced, and of course, all this logic needs to be implemented inside the directive itself, but if these use cases start to arise, the use of this pattern could become a standard.

**Comment by DAreRodz** [comment]
I like the idea of allowing dynamic options for directives. 😄 

Regarding the usage of `ref:` inside the value of a JSON property: it is simply a way to distinguish between different value types, right? It might be better to define a specific property for each type, e.g., `scrollPayload`, `scrollCallback`, etc. That way, you would write the following depending on what you need:

```html
<a data-wp-navigation-link='{ "prefetch": "hover", "scroll": true }'> ... </a>
<a data-wp-navigation-link='{ "prefetch": "hover", "scrollPayload": "selectors.namespace.scrollPayload" }'> ... </a>
<a data-wp-navigation-link='{ "prefetch": "hover", "scrollCallback": "actions.namespace.scrollToSomeElement" }'> ... </a>
```

If you really need to allow passing different types for the same prop, I'd avoid encouraging the usage of special prefixes. A more straightforward way would be to check if the string starts with `actions.`, `selectors.`, etc., which is our current convention for defining store functions.

```html
<a data-wp-navigation-link='{ "prefetch": "hover", "scroll": true }'> ... </a>
<a data-wp-navigation-link='{ "prefetch": "hover", "scroll": "selectors.namespace.scrollPayload" }'> ... </a>
<a data-wp-navigation-link='{ "prefetch": "hover", "scroll": "actions.namespace.scrollToSomeElement" }'> ... </a>
```
